### DISCUSSION TITLE
Custom component pipelines not running in parallel

### DISCUSSION THREAD

**Discussion by chadm8** [openmpf/openmpf#1597] [Q&A] [answered]
Hello,

My team noticed that when we spam create jobs with the OpenMPF provided pipelines, they run very quickly because they run in parallel. We noticed, however, that when we spam create jobs with our custom Deepface pipeline, they run only one at a time, making the process quite slow. Is there any reason for this? We followed the steps for creating custom components and everything runs normally other than the fact that they run sequentially. Any guidance on this would be appreciated.

Thanks

**Comment by jrobble** [comment, marked as answer]
You should be able to set the number of `replicas` for a service entry in your `docker-compose.yml` file like this:

```yml
my-component:
  ...
  deploy:
    mode: replicated
    replicas: 2
```

If you don't have a `deploy` section for your service entry, the default will only create one instance of that service.

If you look in the example `docker-compose.yml` file [here](https://hub.docker.com/r/openmpf/openmpf_workflow_manager), most of the component services extend from a base with 2 replicas.

Please verify that you have the correct number of docker services running. On a single-host deployment (assuming you started the deployment with `docker compose up`) you can run `docker container ls` to list them.

**Comment by brosenberg42** [comment]
We don't treat custom pipelines differently from the built in ones. After you follow Jeff's advice or if you already have multiple replicas running, you can run `docker stats` to see the CPU utilization of each container to verify they are processing jobs. 

Depending on the length of the video, the segment size, and the number of replicas it may appear to be processing the jobs sequentially. When Workflow Manager receives a video job, it creates sub-jobs each of which tells the component to process a certain segment of the video. Say you submit two jobs at approximately the same time each of which gets broken up in to 4 sub-jobs and you have 2 replicas of your component running. In most cases the first job's sub-jobs will enter the component's queue before the second. In that case, the component will completely process the first job before it starts on the second, but the sub-jobs are processed in parallel. If you increase the number of replicas to 5 or greater, then both jobs will run at the same time.
